src/client/pkg/tracing/tracing.go

```diff
@@ -0,0 +1,196 @@
+			key, ok := kvs[i].(string) // common case--key is string
+			if !ok {
+				key = fmt.Sprintf("%v", kvs[i])
+			}
+			val, ok := kvs[i+1].(string) // common case--val is string
+			if !ok {
+				val = fmt.Sprintf("%v", kvs[i+1])
+			}
+			options = append(options, opentracing.Tag{key, val})
+		}
+
+		// construct new span
+		span := opentracing.StartSpan(operation, options...)
+		return span, opentracing.ContextWithSpan(ctx, span)
+	}
+	return nil, ctx
+}
+
+// FinishAnySpan calls span.Finish() if span is not nil. Pairs with
+// AddSpanToAnyExisting
+func FinishAnySpan(span opentracing.Span) {
+	if span != nil {
+		span.Finish()
+	}
+}
+
+// InstallJaegerTracerFromEnv installs a Jaeger client as the opentracing global
+// tracer, relying on environment variables to configure the client
+func InstallJaegerTracerFromEnv() {
+	jaegerOnce.Do(func() {
+		jaegerEndpoint, onUserMachine := os.LookupEnv(jaegerEndpointEnvVar)
+		if !onUserMachine {
+			if host, ok := os.LookupEnv("JAEGER_COLLECTOR_SERVICE_HOST"); ok {
+				port := os.Getenv("JAEGER_COLLECTOR_SERVICE_PORT_JAEGER_COLLECTOR_HTTP")
+				jaegerEndpoint = fmt.Sprintf("%s:%s", host, port)
+			}
+		}
+		if jaegerEndpoint == "" {
+			return // break early -- not using Jaeger
+		}
+
+		// canonicalize jaegerEndpoint as http://<hostport>/api/traces
+		jaegerEndpoint = strings.TrimPrefix(jaegerEndpoint, "http://")
+		jaegerEndpoint = strings.TrimSuffix(jaegerEndpoint, "/api/traces")
+		jaegerEndpoint = fmt.Sprintf("http://%s/api/traces", jaegerEndpoint)
+		cfg := jaegercfg.Configuration{
+			// Configure Jaeger to sample every call, but use the SpanInclusionFunc
+			// addTraceIfTracingEnabled (defined below) to skip sampling every RPC
+			// unless the PACH_ENABLE_TRACING environment variable is set
+			Sampler: &jaegercfg.SamplerConfig{
+				Type:  "const",
+				Param: 1,
+			},
+			Reporter: &jaegercfg.ReporterConfig{
+				LogSpans:            true,
+				BufferFlushInterval: 1 * time.Second,
+				CollectorEndpoint:   jaegerEndpoint,
+			},
+		}
+
+		// configure jaeger logger
+		logger := jaeger.Logger(jaeger.NullLogger)
+		if !onUserMachine {
+			logger = jaeger.StdLogger
+		}
+
+		// Hack: ignore second argument (io.Closer) because the Jaeger
+		// implementation of opentracing.Tracer also implements io.Closer (i.e. the
+		// first and second return values from cfg.New(), here, are two interfaces
+		// that wrap the same underlying type). Instead of storing the second return
+		// value here, just cast the tracer to io.Closer in CloseAndReportTraces()
+		// (below) and call 'Close()' on it there.
+		tracer, _, err := cfg.New(JaegerServiceName, jaegercfg.Logger(logger))
+		if err != nil {
+			panic(fmt.Sprintf("could not install Jaeger tracer: %v", err))
+		}
+		opentracing.SetGlobalTracer(tracer)
+	})
+}
+
+// addTraceIfTracingEnabled is an otgrpc interceptor option that propagates
+// existing traces, but won't start any new ones
+var addTraceIfTracingEnabled otgrpc.SpanInclusionFunc = func(
```

Comment:
Why is this declared as a `var` rather than as a normal func? I think if you implement the right signature you don't need to declare it as a `SpanInclusionFunc`. It might require a cast though.

Reply:
Done